container/go/cmd/zipper/zipper.go

```diff
@@ -0,0 +1,81 @@
+// Copyright 2020 The Bazel Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+////////////////////////////////////////////////
+// This package performs gzip operations.
+
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"flag"
+	"io/ioutil"
+	"log"
+	"os"
+)
+
+var (
+	src        = flag.String("src", "", "The source location of the file to zip/unzip.")
+	dst        = flag.String("dst", "", "The destination location of the file, after zip/unzip.")
+	decompress = flag.Bool("decompress", false, "If true, perform gunzip. If false, gzip.")
+	fast       = flag.Bool("fast", false, "If true, use the fastest compression level.")
+)
+
+func main() {
+	flag.Parse()
+
+	if *src == "" {
+		log.Fatalln("Required option -src was not specified.")
+	}
+
+	if *dst == "" {
+		log.Fatalln("Required option -dst was not specified.")
+	}
+
+	dat, err := ioutil.ReadFile(*src)
+	if err != nil {
+		log.Fatalf("Unable to read input file: %v", err)
+	}
+	var buf bytes.Buffer
```

Comment:
Suggest using io.Copy (https://pkg.go.dev/io?tab=doc#Copy) to directly copy from a buffer pointing to the source file to a buffer pointing to the destination file. This avoids loading the entire src into memory.

When decompressing, the src would be a gzip.NewReader(dat), dst would be os.Create (https://pkg.go.dev/os?tab=doc#Create). When compressing, src would be dat and dst would be gzip.NewWriterLevel wrapping a os.Create.

Finally use io.Copy (https://pkg.go.dev/io?tab=doc#Copy) to copy from the source to destination buffer. If you declare io.Reader (https://pkg.go.dev/io?tab=doc#Reader) and io.Writer interface variables outside the if conditions, you can have one io.Copy at the end that handles both cases.

Reply:
Done.